Add fast Fewest-Pending-Request load balancer

kriskowal opened the pull request:
This change introduces a novel "circus" peer list.

The primary premise of the circus peer list is that a production peer list will typically only have two classes of peer: those that have a certain number of pending requests, and those that have one more than that.

Making this assumption, we can use a pair of circular linked lists to model the highly and lightly loaded peer classes and move peers between them only when they begin and end requests.

When the list of lightly loaded peers is empty, we can swap the lists.  This would be more likely necessary when aggregate load is increasing.

This is a three ring circus.  Each ring is a circular doubly linked list with a head node.  One ring for unavailable peers, those waiting for an open connection to become available.  The second and third ring are for peers with more or fewer concurrent requests.  A fourth ring tracks unused nodes.

The second premise is that the peer list will never need to contain more than 256 peers.  Although it is possible to have more, beyond that point, it is okay to sample.

Making this assumption we can use a fixed allocation of 256 nodes shared by the four circular doubly linked lists and their head nodes.  Because of the size of the allocation, the internal references of the list can just be single bytes.

A third idea is to model the internal tables as columns.  In general, this has the benefit of separating some concerns and potentially reducing CPU cache misses.  For example, the linked lists do not need to contain their corresponding data type and we dodge a genericity bullet.  We use the linked lists to track allocation and the topology of rings, and the coresponding indices in other arrays to track peers and subscribers.

Preliminary measurements suggest that this algorithm can choose peers twice as fast as the pending heap for lists on the order of 64 peers.

AlexAPB commented:
I still need to take a closer look, but let's put this under an `x/` package until we can do more thorough analysis / validate with a customer on this one.

kriskowal commented:
I’ve moved into x/peer/circus and added an internal/benchmark/peerchoosefinish that dumps a spreadsheet out.

![image](https://user-images.githubusercontent.com/60294/74474029-847c0e80-4e59-11ea-8c06-65001afc10f2.png)

AlexAPB commented on peer/x/circus/list.go:
```diff
@@ -0,0 +1,368 @@
+		l.nodes.shift(index, _no)
+	}
+}
+
```
May be worth indicating that the below code is lifted from the abstract peer list

AlexAPB commented on peer/x/circus/list.go:
```diff
@@ -0,0 +1,368 @@
+		// 	panic(fmt.Sprintf("add already added: %s", addr))
+		// }
+		l.index[addr] = 0
+		// The maximum capacity of this peer list is 252 peers.
```
Why do we have this cap? This should be called out in the top-level package docs.

AlexAPB commented on peer/x/circus/nodes_test.go:
```diff
@@ -0,0 +1,130 @@
+			nodes.shift(nodes[_free].next, _lo)
+			count++
+		}
+		assert.Equal(t, size-4, count)
```
Ah, capped at 252. Can we just make this a top-level var `var _max = _size - 4`

kriskowal commented:
Closing because science reveals no improvement over round robin in simulation. 🔬 

Simulation https://github.com/yarpc/yarpc-go/pull/1893

![image](https://user-images.githubusercontent.com/60294/75073625-b0793e80-54ae-11ea-9632-3771706c193c.png)
